File: backend/src/handlers/usuarios.rs
use axum::extract::{Path, Query, State};
use axum::routing::get;
use axum::{Extension, Json, Router};
use uuid::Uuid;

use crate::auth::models::Claims;
use crate::db::AppState;
use crate::dto::usuario::{
    CreateUsuario, ResetPasswordRequest, UpdateUsuario, UsuarioQuery, UsuarioResponse,
};
use crate::errors::AppError;
use crate::services::usuario_service;

async fn listar(
    State(state): State<AppState>,
    Extension(claims): Extension<Claims>,
    Query(params): Query<UsuarioQuery>,
) -> Result<Json<Vec<UsuarioResponse>>, AppError> {
    crate::auth::middleware::require_role(&["admin"])(&claims)?;
    let usuarios = usuario_service::listar(&state.pool, params).await?;
    Ok(Json(usuarios))
}

async fn obtener(
    State(state): State<AppState>,
    Extension(claims): Extension<Claims>,
    Path(id): Path<Uuid>,
) -> Result<Json<UsuarioResponse>, AppError> {
    crate::auth::middleware::require_role(&["admin"])(&claims)?;
    let usuario = usuario_service::obtener(&state.pool, id).await?;
    Ok(Json(usuario))
}

async fn crear(
    State(state): State<AppState>,
    Extension(claims): Extension<Claims>,
    Json(req): Json<CreateUsuario>,
) -> Result<(axum::http::StatusCode, Json<UsuarioResponse>), AppError> {
    crate::auth::middleware::require_role(&["admin"])(&claims)?;
    let usuario = usuario_service::crear(&state.pool, req, claims.sub).await?;
    Ok((axum::http::StatusCode::CREATED, Json(usuario)))
}

async fn actualizar(
    State(state): State<AppState>,
    Extension(claims): Extension<Claims>,
    Path(id): Path<Uuid>,
    Json(req): Json<UpdateUsuario>,
) -> Result<Json<UsuarioResponse>, AppError> {
    crate::auth::middleware::require_role(&["admin"])(&claims)?;
    let usuario = usuario_service::actualizar(&state.pool, id, req, claims.sub).await?;
    Ok(Json(usuario))
}

async fn eliminar(
    State(state): State<AppState>,
    Extension(claims): Extension<Claims>,
    Path(id): Path<Uuid>,
) -> Result<axum::http::StatusCode, AppError> {
    crate::auth::middleware::require_role(&["admin"])(&claims)?;
    usuario_service::eliminar(&state.pool, id, claims.sub).await?;
    Ok(axum::http::StatusCode::NO_CONTENT)
}

async fn reset_password(
    State(state): State<AppState>,
    Extension(claims): Extension<Claims>,
    Path(id): Path<Uuid>,
    Json(req): Json<ResetPasswordRequest>,
) -> Result<axum::http::StatusCode, AppError> {
    crate::auth::middleware::require_role(&["admin"])(&claims)?;
    usuario_service::reset_password(&state.pool, id, req.password_nueva, claims.sub).await?;
    Ok(axum::http::StatusCode::NO_CONTENT)
}

pub fn routes() -> Router<AppState> {
    Router::new()
        .route("/", get(listar).post(crear))
        .route("/{id}", get(obtener).put(actualizar).delete(eliminar))
        .route("/{id}/reset-password", axum::routing::post(reset_password))
}

File: frontend/src/api/usuarios.ts
// Dominio: usuarios
import api from '@/lib/api'
import type {
  UsuarioResponse,
  CreateUsuario,
  UpdateUsuario,
  UsuarioQuery,
  ResetPasswordRequest,
} from '@/types/generated'

// ─── Funciones ────────────────────────────────────────────────────────────────

/** GET /usuarios — Listar usuarios con filtros opcionales */
export async function listarUsuarios(params?: Partial<UsuarioQuery>): Promise<UsuarioResponse[]> {
  const { data } = await api.get<UsuarioResponse[]>('/usuarios', { params })
  return data
}

/** GET /usuarios/:id — Obtener detalle de un usuario */
export async function detalleUsuario(id: string): Promise<UsuarioResponse> {
  const { data } = await api.get<UsuarioResponse>(`/usuarios/${id}`)
  return data
}

/** POST /usuarios — Crear nuevo usuario */
export async function crearUsuario(payload: CreateUsuario): Promise<UsuarioResponse> {
  const { data } = await api.post<UsuarioResponse>('/usuarios', payload)
  return data
}

/** PUT /usuarios/:id — Actualizar usuario */
export async function actualizarUsuario(id: string, payload: UpdateUsuario): Promise<UsuarioResponse> {
  const { data } = await api.put<UsuarioResponse>(`/usuarios/${id}`, payload)
  return data
}

/** DELETE /usuarios/:id — Desactivar usuario (soft delete) */
export async function eliminarUsuario(id: string): Promise<void> {
  await api.delete(`/usuarios/${id}`)
}

/** POST /usuarios/:id/reset-password — Resetear contraseña de un usuario */
export async function resetPassword(id: string, payload: ResetPasswordRequest): Promise<void> {
  await api.post(`/usuarios/${id}/reset-password`, payload)
}
